### PR TITLE
Add in call backs for drag/resize calls

### PR DESCRIPTION
**Pull request by browniefed**
Thoughts on adding callbacks for this stuff?
I just basically passed through most information since I wasn't sure exactly what people were going to want to utilize.

### CODE REVIEW

File: build/GridItem.js [hunk 1]
```diff
@@ -70,7 +70,7 @@ var GridItem = React.createClass({
     handle: React.PropTypes.string
   },
 
-  getDefaultProps: function () {
+  getDefaultProps: function getDefaultProps() {
```
Comment (STRML) [inline build/GridItem.js]: Which version of 6to5 are you using? I wonder why the build came out this way for you; this is definitely better for stack traces.

File: lib/ReactGridLayout.jsx [hunk 2]
```diff
@@ -57,7 +57,16 @@ var ReactGridLayout = React.createClass({
...
     onLayoutChange: React.PropTypes.func,
-
+    // Calls when drag starts, allows for canceling/restricting drag
+    onDragStart: React.PropTypes.func,
+    // Calls on each drag movement, allows for modifying current drag layout, or preventing layout modificaiton
+    onDrag: React.PropTypes.func,
+    // Cals when drag is complete
+    onDragStop: React.PropTypes.func,
+    // Calls when resize starts, allows to cancel/restrict resize
+    onResize: React.PropTypes.func,
```
Comment (STRML) [inline lib/ReactGridLayout.jsx]: I suppose we should add an onResizeStart to be consistent?

File: lib/ReactGridLayout.jsx [hunk 3]
```diff
@@ -152,11 +166,33 @@ var ReactGridLayout = React.createClass({
...
+   * @param {Number} i Index of the child
+   * @param {Number} x X position of the move
+   * @param {Number} y Y position of the move
+   * @param {Element} element The current dragging DOM element
+   * @param {Object} position Drag information
+   */
+  onDragStart(i, x, y, {e, element, position}) {
+    var layout = this.state.layout;
+    var l = utils.getLayoutItem(layout, i);
+    if (this.props.onDragStart(layout, this.state.layout, l, null, {e, element, position}) === false) {
```
Comment (STRML) [inline lib/ReactGridLayout.jsx]: I like making it cancelable via the callback; any ideas how this is done in other libraries? My instinct is to check if the event has been canceled (e.g. e.preventDefault()) but that doesn't quite make as much sense in React.

Comment (STRML) [inline lib/ReactGridLayout.jsx]: Of course if we did that we'd have to be consistent among all the other callbacks as well. With this code as-is the placeholder would quickly be replaced by the drag handler below.

File: build/GridItem.js (hunk 1 above)
Comment (browniefed) [inline build/GridItem.js]: I just did an npm install so whatever the package.json says

File: lib/ReactGridLayout.jsx (hunk 2 above)
Comment (browniefed) [inline lib/ReactGridLayout.jsx]: I'll add it

File: lib/ReactGridLayout.jsx (hunk 3 above)
Comment (browniefed) [inline lib/ReactGridLayout.jsx]: Yeah, I think e.preventDefault would be the way to go since I'm not necessarily a fan of returning false, and as of React .12 you must call preventDefault to stop.
It looks like we could do it with mouseDown event since that is what react-draggable is using, however we may need to mod the handleStartDrag to actually do the preventDefault check.

File: build/GridItem.js (hunk 1 above)
Comment (STRML) [inline build/GridItem.js]: Gotcha. I had an old version in npm cache. Could you rebase to master so we can squash the diffs?

File: lib/ReactGridLayout.jsx (hunk 3 above)
Comment (STRML) [inline lib/ReactGridLayout.jsx]: That's fine by me, I'll add that to my fork of draggable.

File: build/GridItem.js (hunk 1 above)
Comment (browniefed) [inline build/GridItem.js]: Will do, I'm at work now so I'll handle all these sometime later tonight.

Comment (STRML) [inline build/GridItem.js]: Cool. Thanks a lot for the contribution, I appreciate it.

**Comment by browniefed** [comment]
Rebased, added onResizeStart, add README documentation.
Only thing this will depend on is react-draggable and react-resizable stopping drag/resize when `preventDefault` gets called.

**Comment by browniefed** [comment]
Hey just curious if you had an idea of when these could get merged?

**Comment by STRML** [comment]
Hey Jason - sorry, chasing down a lot of things on a lot of projects
right now. Should be merged soon, in the meantime I would recommend
keeping this rebased on master and using a github url in your project
until I can finish reviewing.

**Comment by browniefed** [comment]
@STRML no problem, thanks

File: lib/ReactGridLayout.jsx [hunk 4]
```diff
@@ -201,13 +258,20 @@ var ReactGridLayout = React.createClass({
     var activeDrag = {
       w: w, h: h, x: l.x, y: l.y, placeholder: true, i: i
     };
+
+    activeDrag = this.props.onResize(layout, layout, l, activeDrag, {e, element, size}) || activeDrag;
```
Comment (STRML) [inline lib/ReactGridLayout.jsx]: In general I don't really like the assignment here; I don't see any demand for it and it is confusing - on dragStart and resizeStart, returning a value does nothing, on drag, you return the layout, on resize, you return the resizing item, and on dragStop and resizeStop you return the layout again.

Do you have a particular need for that functionality? Otherwise let's just remove the assignment altogether.

File: README.md [hunk 5]
```diff
@@ -181,6 +181,18 @@ listenToWindowResize: React.PropTypes.bool,
 // Callback so you can save the layout.
 // Calls back with (currentLayout, allLayouts). allLayouts are keyed by breakpoint.
 onLayoutChange: React.PropTypes.func,
+// Calls when drag starts, allows for canceling/restricting drag    
+onDragStart: React.PropTypes.func,
+// Calls on each drag movement, allows for modifying current drag layout, or preventing layout modificaiton
+onDrag: React.PropTypes.func,
+// Calls when drag is complete
+onDragStop: React.PropTypes.func,
+//Calls when resize starts, allows for canceling
```
Comment (STRML) [inline README.md]: Could you please fix the spacing here and note the callback signature on each one both here and in the source?

For example: 

"Calls when resize starts. Callback is of the signature `(newLayout, oldLayout, layoutItem, {e, element, size}`."

Modify the callback signature to `{e, element, position}` for drags.

File: lib/ReactGridLayout.jsx (hunk 4 above)
Comment (browniefed) [inline lib/ReactGridLayout.jsx]: Well onResizeStart/onDragStart you ideally would be canceling with preventDefault so no return necessary.
Ideally if there were a way to set both full layout/active drag layout items that would be ideal.
I'm attempting to implement a lot of restrictions around dragging/resizing  so controlling the full layout/actively dragging/resizing item is what I need.

I realize I can modify the layout and activeDrag by just simple object reference but that's just bad. Besides calling a setState of ReactGridLayout and passing down new props layout during the drag (which would probably be very inefficient and cause some weird UI states temporarily) I'm not sure what else is the solution here.

Basically how would you allow the user to control the positioning of the entire layout/actively dragging item while it's happening?

**Comment by browniefed** [comment]
@STRML hey I just decided to remove the assignment stuff for now, so I've rebased and squashed, fixed the documentation text. I think it should be good to go.

**Comment by STRML** [comment]
Perfect. This is simple, we can worry about the assignment stuff later on.

**Comment by STRML** [comment]
`0.7.1`
